### PR TITLE
Add a longer chase to get around CMC prefetcher

### PR DESCRIPTION
**Pull request by puj2**

### CODE REVIEW

File: permutation.c [hunk 1]
```diff
@@ -182,3 +184,100 @@ void *generate_chase(const struct generate_chase_common_args *args,
...
+    gen_permutation(tlb_perm, nr_tlb_groups, 0);
+
+    for (i = 0; i < nr_tlb_groups; ++i) {
+      gen_permutation(&perm[j * nr_elts + i * nr_elts_per_tlb], nr_elts_per_tlb,
+                    base + tlb_perm[i] * nr_elts_per_tlb);
+    }
+    free(tlb_perm);
+
+    dassert(is_a_permutation(perm, nr_elts));
+    if (verbosity > 1) printf("generating inverse permtuation\n");
```
Comment (seranian) [inline permutation.c:233]: typo on permutation

File: permutation.c [hunk 2]
```diff
@@ -182,3 +184,100 @@ void *generate_chase(const struct generate_chase_common_args *args,
...
+  size_t base;
+  perm_t *perm_inverse;
+  size_t mixer_scale = stride / nr_mixer_indices;
+
+  if (verbosity > 1)
+    printf("generating permutation of %zu elements (in %zu TLB groups)\n",
+           nr_elts, nr_tlb_groups);
+
+  perm = malloc(nr_iteration * nr_elts * sizeof(*perm));
+
```
Comment (seranian) [inline permutation.c:218]: no checking on return value of malloc()

File: permutation.c [hunk 3]
```diff
@@ -182,3 +184,100 @@ void *generate_chase(const struct generate_chase_common_args *args,
...
+  // Generate nr_iteration number of permutations.
+  for (j = 0; j < nr_iteration; j++) {
+    base = j * nr_elts;
+
+    tlb_perm = malloc(nr_tlb_groups * sizeof(*tlb_perm));
```
Comment (seranian) [inline permutation.c:223]: no checking on return value of malloc()

File: permutation.c [hunk 4]
```diff
@@ -182,3 +184,100 @@ void *generate_chase(const struct generate_chase_common_args *args,
...
+    }
+    free(tlb_perm);
+
+    dassert(is_a_permutation(perm, nr_elts));
+    if (verbosity > 1) printf("generating inverse permtuation\n");
+  }
+
+  dassert(is_a_permutation(perm, nr_iteration * nr_elts));
+
+  perm_inverse = malloc(nr_iteration * nr_elts * sizeof(*perm_inverse));
```
Comment (seranian) [inline permutation.c:238]: no checking of return value of malloc()

**Review by seranian — COMMENTED** [review]
some comments inlined

File: permutation.c (hunk 4 above)
Comment (puj2) [reply permutation.c:250]: Added checking

File: permutation.c (hunk 3 above)
Comment (puj2) [reply permutation.c:228]: Added checking

File: permutation.c (hunk 2 above)
Comment (puj2) [reply permutation.c:223]: Added checking

File: permutation.c (hunk 1 above)
Comment (puj2) [reply permutation.c]: Fixed

**Pull request merged**